### ISSUE TITLE
 How to use brave-opentracing API to let Zipkin to display Server Receive and  Server Send？

### ISSUE THREAD

**Issue by tomgithub2016** [openzipkin/brave-opentracing#18]
In my project,I need to use （opentacing-api+brave-opentracing+Zipkin） to tracing.
 I use brave-opentracing API  start(),finish() to create Span and post ,But In Zipkin only to display Client Send and Client Receive datas ,do not display Server Receive and Server Send.

My questions is  How to use brave-opentracing API to let Zipkin to display Server Receive and  Server Send?

Demo,  There are two micro services,ServiceA and ServiceB
ServiceA  remote procedure call ServiceB，via brower to call ServiceA and ServiceA will  remote call serviceB，and ServiceA、ServiceB span finish();
But in zipkin,only to display Client Send and Client Receive,
In this demo, How to display  Server Receive and  Server Send by using brave-opentracing API ?

In ServiceA code
```
@RequestMapping("ServiceA")
	public String serviceA() throws InterruptedException, IOException
	{
		//BraveTracer  
		//root span
 		Span span0 = tracer.buildSpan("ServiceA").start();
		Map<String, String> map = new HashMap<String, String>();
		tracer.inject(span0.context(), Format.Builtin.HTTP_HEADERS, new TextMapInjectAdapter(map));
 		//build HttpHeader
		HttpHeaders header = new HttpHeaders();
		header.setAll(map);
		//remote invoke
 		String rs = restTemplate.postForEntity("http://localhost:9393/ServiceB", new HttpEntity(header), String.class)
				.getBody();
 		//post
 		span0.finish();
 		return "";
	}
```

In ServiceB code

```

	@RequestMapping("ServiceB")
	public String serviceB() throws InterruptedException, IOException
	{
 
		Enumeration<String> e = request.getHeaderNames();
		Map<String, String> spanMap = new HashMap<String, String>();
 
		while (e.hasMoreElements())
		{
			// add the names of the request headers into the spanMap  
			String key = e.nextElement();
			String value = request.getHeader(key);
			spanMap.put(key, value);
 		}
		Span span1 = tracer.buildSpan("ServiceB")
				.asChildOf(tracer.extract(Format.Builtin.HTTP_HEADERS, new TextMapExtractAdapter(spanMap))).start();
		
		
		//demo doSomethings
		Thread.sleep(3000);

		//ServiceB post
 		span1.finish();

		return "test";
	} 
```

**Comment by adriancole** [comment]
I think what should happen is that when the tag "span.kind" is sent, on the
brave side, do span.kind(SERVER) in this case.

That's not currently wired up, so we'd need a PR with tests.

**Comment by tomgithub2016** [comment]
@adriancole     
 I have solved this problem,　According to the code,
To add additional code in client Endpoint

```java
   braveSpan.annotate(Constants.CLIENT_SEND);
```
and add code in server Endpoint
```java
   braveSpan.annotate(Constants.SERVER_RECV);
```
Via brave-opentracing api，you can custom design  annotate to  zipkin.
for example
```java
Span span1 = opentracing.buildSpan("Span1").withTag(Constants.LOCAL_COMPONENT, "client").start();
brave.Span braveSpan = ((BraveSpan) span1).unwrap();
braveSpan.annotate(start, "show");
span0.finish();

```
Zipkin will display CS CR  SR SS，

thanks！

**Comment by adriancole** [comment]
yep that will work. slightly less work would be to ((BraveSpan) span1).
unwrap().setKind(CLIENT)

less because you only need to do it once (as opposed to annotate twice)

**Comment by adriancole** [comment]
This isn't complete, as the approach you mentioned is not a complete fix (as you'll notice in #19)

What needs to happen is what I mentioned here https://github.com/openzipkin/brave-opentracing/issues/18#issuecomment-279658289

it needs to have tests so that we don't keep getting similar issues on the same topic of RPC spans

**Comment by tomgithub2016** [comment]
I have solved this problem, the root reason of the problem is ServiceA request To ServiceB CR、CS、SS、SR spanId must be same; In this conodition Zipkin-dependencies can run succssful.

**Comment by adriancole** [comment]
nice here's more on that topic, if interested
https://github.com/openzipkin/b3-propagation

**Comment by adriancole** [comment]
fixed by @pavolloffay here #24
